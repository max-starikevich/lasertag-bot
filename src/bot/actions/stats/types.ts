export enum GameResult {
  RED = 'RED',
  BLUE = 'BLUE',
  DRAW = 'DRAW'
}

export interface GameData {
  id: string
  date: number
  red: string[]
  blue: string[]
  telegramUserId: number
}
