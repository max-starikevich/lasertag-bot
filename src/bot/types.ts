import { NarrowedContext } from 'telegraf'
import { ChatMember, Message, Update } from 'telegraf/typings/core/types/typegram'

import { Player } from '$/game/player/types'

import { BaseLogger } from '$/logger/types'
import { Locales, TranslationFunctions } from '$/lang/i18n-types'
import { GameStorage, GameStore } from '$/game/storage/types'
import { CustomContext } from './CustomContext'

export interface GameContext extends CustomContext {
  storage: GameStorage
  store: GameStore
  logger: BaseLogger

  isAdminInHomeChat: boolean
  isCreatorOfHomeChat: boolean

  isPrivateChat: boolean

  memberStatus: ChatMember['status']

  lang: TranslationFunctions
  locale: Locales

  currentPlayer: Player | undefined
  players: Player[]
}

export type CommandContext = NarrowedContext<GameContext, Update.MessageUpdate<Message.TextMessage>>
export type CommandHandler = (ctx: CommandContext) => Promise<any>

export interface Command {
  name: string
  handler: CommandHandler
  description: (lang: TranslationFunctions) => string
  showInMenu: boolean
}

export type ActionHandler = (ctx: NarrowedContext<GameContext & {
  match: RegExpExecArray
}, Update.CallbackQueryUpdate>) => Promise<any>

export type ActionInitializer = (ctx: CommandContext) => Promise<any>

export interface Action {
  mapping: { [key: string]: ActionHandler }
}
