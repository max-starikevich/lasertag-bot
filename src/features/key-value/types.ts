export interface IKeyValueStore {
  get: <T>(keys: string[]) => Promise<Array<KeyValue<T>>>
  set: <T>(objects: Array<KeyValue<T>>) => Promise<void>
  delete: (keys: string[]) => Promise<void>
}

export interface KeyValue<T> {
  key: string
  value: T | null
}
