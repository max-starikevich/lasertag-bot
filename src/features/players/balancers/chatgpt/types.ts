export interface AiBalanceOutput {
  team1: string[]
  team2: string[]
}

export const balanceOutputExample: AiBalanceOutput = {
  team1: [
    'Emma',
    'Liam',
    'Olivia',
    'Noah',
    'Ava',
    'Ethan',
    'Isabella',
    'Mason'
  ],
  team2: [
    'Sophia',
    'Jacob',
    'Mia',
    'William',
    'Amelia',
    'James',
    'Harper'
  ]
}

export interface PlayerWithSkills {
  Name: string
  [skillName: string]: string
}

export interface ISkillsRepository {
  find: (keys: string[]) => Promise<PlayerWithSkills[]>
}
