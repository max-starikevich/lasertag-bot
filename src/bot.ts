import { Telegraf } from 'telegraf';

import { BotContext } from '@/types';
import { setBotCommands } from '@/commands';
import config from '@/config';
import { getSpreadsheetDocument } from '@/sheets';

export const prepareBot = async () => {
  const bot = new Telegraf<BotContext>(config.BOT_TOKEN);
  bot.context.document = getSpreadsheetDocument();

  await setBotCommands(bot);

  return bot;
};
