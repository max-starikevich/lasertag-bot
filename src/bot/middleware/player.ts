import { BotMiddleware } from '.'

import { isLocaleName } from '$/lang/i18n-custom'
import { Locales } from '$/lang/i18n-types'
import L from '$/lang/i18n-node'

export const playerMiddleware: BotMiddleware = async (ctx, next) => {
  if (ctx.from == null) {
    throw new Error('Missing "ctx.from"')
  }

  const { game } = ctx

  const players = await game.getPlayers()

  const currentPlayer = players.find(({ telegramUserId }) => telegramUserId !== undefined && telegramUserId === ctx.from?.id)

  if (currentPlayer === undefined) {
    return await next()
  }

  ctx.currentPlayer = currentPlayer

  if (isLocaleName(currentPlayer.locale) === true) {
    ctx.locale = currentPlayer.locale as Locales
    ctx.lang = L[ctx.locale]
  }

  return await next()
}
