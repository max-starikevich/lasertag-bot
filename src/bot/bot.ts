import { Telegraf } from 'telegraf'

import { GameContext } from '$/bot/types'
import { commands } from '$/bot/commands'
import { setBotActions, setBotMiddlewares } from '$/bot/middleware'

import { GameStorage, GameStore } from '$/game/storage/types'
import L from '$/lang/i18n-node'
import { defaultLocale } from '$/lang/i18n-custom'

import { errorMiddleware } from './middleware/error'
import { CustomContext } from './CustomContext'
import ApiClient from 'telegraf/typings/core/network/client'

export const commandsInMenu = commands.filter(
  ({ showInMenu }) => showInMenu
)

interface InitBotParams {
  storage: GameStorage
  store: GameStore
  token: string
  telegram?: Partial<ApiClient.Options>
}

export const initBot = async ({ token, telegram, storage, store }: InitBotParams): Promise<Telegraf<GameContext>> => {
  const bot = new Telegraf<GameContext>(token, {
    // @ts-expect-error
    contextType: CustomContext,
    telegram
  })

  bot.context.storage = storage
  bot.context.store = store
  bot.context.players = []

  // will be overriden in the access middleware
  bot.context.isAdminInHomeChat = false
  bot.context.isCreatorOfHomeChat = false
  bot.context.isPrivateChat = false

  bot.context.lang = L[defaultLocale]
  bot.context.locale = defaultLocale

  setBotMiddlewares(bot)
  setBotActions(bot)

  bot.catch(errorMiddleware)

  return bot
}
