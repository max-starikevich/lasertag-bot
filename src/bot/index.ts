import { Telegraf } from 'telegraf'

import config from '$/config'
import { checkEnvironment } from '$/config/check'

import { GameContext } from '$/bot/types'
import { commands } from '$/bot/commands'
import { setBotActions, setBotMiddlewares } from '$/bot/middleware'

import { Game } from '$/game/Game'
import { GoogleTableGameStorage } from '$/game/storage/google-table/GoogleTableGameStorage'
import { reportException } from '$/errors'
import L from '$/lang/i18n-node'
import { defaultLocale } from '$/lang/i18n-custom'

import { errorMiddleware } from './middleware/error'

export const commandsInMenu = commands.filter(
  ({ showInMenu }) => showInMenu
)

export const initBot = async (): Promise<Telegraf<GameContext>> => {
  await checkEnvironment()

  const storage = new GoogleTableGameStorage({
    email: config.GOOGLE_SERVICE_ACCOUNT_EMAIL,
    privateKey: config.GOOGLE_PRIVATE_KEY,
    players: {
      docId: config.PLAYERS_DOC_ID,
      sheetsId: config.PLAYERS_SHEETS_ID
    },
    game: {
      docId: config.GAME_DOC_ID,
      sheetsId: config.GAME_SHEETS_ID
    },
    links: {
      docId: config.LINKS_DOC_ID,
      sheetsId: config.LINKS_SHEETS_ID
    },
    stats: {
      docId: config.STATS_DOC_ID,
      sheetsId: config.STATS_SHEETS_ID
    },
    enroll: {
      docId: config.ENROLL_DOC_ID,
      sheetsId: config.ENROLL_SHEETS_ID,
      ranges: {
        names: config.ENROLL_NAMES_RANGE,
        count: config.ENROLL_COUNT_RANGE,
        rent: config.ENROLL_RENT_RANGE,
        comment: config.ENROLL_COMMENT_RANGE
      }
    }
  })

  const game = new Game({ storage })
  const bot = new Telegraf<GameContext>(config.BOT_TOKEN)

  bot.context.game = game

  // will be overriden in the access middleware
  bot.context.isAdmin = false
  bot.context.isCreator = false
  bot.context.isGroupChat = false
  bot.context.isPrivateChat = false

  bot.context.lang = L[defaultLocale]
  bot.context.locale = defaultLocale

  setBotMiddlewares(bot)
  setBotActions(bot)

  bot.catch(errorMiddleware)

  process.on('uncaughtException', e => reportException(e))
  process.on('unhandledRejection', e => reportException(e))

  return bot
}
