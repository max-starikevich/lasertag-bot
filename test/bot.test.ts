import { Telegraf } from 'telegraf'
import TelegramServer from 'telegram-test-api'
import { TelegramClient } from 'telegram-test-api/lib/modules/telegramClient'

import { initBot } from '$/bot/bot'
import { GameContext } from '$/bot/types'
import { GameStorage, GameStore, StoreData } from '$/game/storage/types'
import { Player, GameStatsData } from '../src/game/player/types'
import { GameLocation, GameLink } from '../src/game/types'

describe('Telegraf bot', () => {
  const tgServer = new TelegramServer({
    port: 8080
  })

  const token = 'test123'

  let bot: Telegraf<GameContext>
  let client: TelegramClient

  beforeAll(async () => {
    await tgServer.start()

    client = tgServer.getClient(token)

    const store: GameStore = {
      get: async function <T>(keys: string[]): Promise<Array<StoreData<T>>> {
        throw new Error('Function not implemented.')
      },
      set: async function <T>(data: Array<StoreData<T>>): Promise<void> {
        throw new Error('Function not implemented.')
      },
      delete: async function (keys: string[]): Promise<void> {
        throw new Error('Function not implemented.')
      }
    }

    const storage: GameStorage = {
      getPlayers: async function (): Promise<Player[]> {
        throw new Error('Function not implemented.')
      },
      getLocations: async function (): Promise<GameLocation[]> {
        throw new Error('Function not implemented.')
      },
      getLinks: async function (): Promise<GameLink[]> {
        throw new Error('Function not implemented.')
      },
      savePlayer: async function (name: string, fields: Partial<Player>): Promise<void> {
        throw new Error('Function not implemented.')
      },
      saveStats: async function (statsData: GameStatsData): Promise<void> {
        throw new Error('Function not implemented.')
      },
      getStatsTimezone: function (): string {
        throw new Error('Function not implemented.')
      }
    }

    bot = await initBot({
      token,
      storage,
      store,
      telegram: {
        apiRoot: tgServer.config.apiURL
      }
    })

    await bot.launch()
  })

  afterAll(async () => {
    await tgServer.stop()
  })

  describe('/about', () => {
    it('should properly get the information', async () => {
      const message = client.makeMessage('/help')
      await client.sendMessage(message)
    })
  })
})
