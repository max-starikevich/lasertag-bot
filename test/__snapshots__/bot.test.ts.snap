// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Telegraf bot /about should properly get the information 1`] = `"test test"`;
